pkg/rtc/subscriptionmanager_test.go

```diff
@@ -0,0 +1,386 @@
+		}
+		sm.params.OnSubcriptionError = func(trackID livekit.TrackID) {
+			failed.Store(true)
+		}
+		numParticipantSubscribed := atomic.Int32{}
+		numParticipantUnsubscribed := atomic.Int32{}
+		sm.OnSubscribeStatusChanged(func(pubID livekit.ParticipantID, subscribed bool) {
+			if subscribed {
+				numParticipantSubscribed.Add(1)
+			} else {
+				numParticipantUnsubscribed.Add(1)
+			}
+		})
+
+		sm.SubscribeToTrack("track", "pub", "pubID")
+		s := sm.subscriptions["track"]
+		require.True(t, s.isDesired())
+		require.Nil(t, s.getSubscribedTrack())
+		require.Eventually(t, func() bool {
+			return subCount.Load() == 1
+		}, subSettleTimeout, subCheckInterval, "track was not subscribed")
+
+		require.Len(t, sm.GetSubscribedTracks(), 1)
+		require.Len(t, sm.GetSubscribedParticipants(), 1)
+		require.Equal(t, "pubID", string(sm.GetSubscribedParticipants()[0]))
+
+		// ensure telemetry events are sent
+		tm := sm.params.Telemetry.(*telemetryfakes.FakeTelemetryService)
+		require.Equal(t, 1, tm.TrackSubscribeRequestedCallCount())
+
+		// ensure bound
+		setTestSubscribedTrackBound(t, s.getSubscribedTrack())
+
+		require.Eventually(t, func() bool {
+			return !s.needsBind()
+		}, subSettleTimeout, subCheckInterval, "track was not bound")
+
+		// telemetry event should have been sent
+		require.Equal(t, 1, tm.TrackSubscribedCallCount())
+
+		time.Sleep(notFoundTimeout)
+		require.False(t, failed.Load())
+
+		// ensure its resilience after being closed
+		setTestSubscribedTrackClosed(t, s.getSubscribedTrack(), false)
+		require.True(t, s.needsSubscribe())
+
+		require.Eventually(t, func() bool {
+			return s.isDesired() && !s.needsSubscribe()
+		}, subSettleTimeout, subCheckInterval, "track was not resubscribed")
+
+		// was subscribed twice, unsubscribed once (due to close)
+		require.Equal(t, int32(2), numParticipantSubscribed.Load())
+		require.Equal(t, int32(1), numParticipantUnsubscribed.Load())
+	})
+
+	t.Run("no track permission", func(t *testing.T) {
+		sm := newTestSubscriptionManager(t)
+		defer sm.Close(false)
+		resolver := newTestResolver(false, nil)
+		sm.params.TrackResolver = resolver.Resolve
+		failed := atomic.Bool{}
+		sm.params.OnSubcriptionError = func(trackID livekit.TrackID) {
+			failed.Store(true)
+		}
+
+		sm.SubscribeToTrack("track", "pub", "pubID")
+		s := sm.subscriptions["track"]
+		require.Eventually(t, func() bool {
+			return !s.getHasPermission()
+		}, subSettleTimeout, subCheckInterval, "should not have permission to subscribe")
+
+		time.Sleep(subscriptionTimeout)
+
+		// should not have called failed callbacks, isDesired remains unchanged
+		require.True(t, s.isDesired())
+		require.False(t, failed.Load())
+		require.True(t, s.needsSubscribe())
+		require.Len(t, sm.GetSubscribedTracks(), 0)
+
+		// trackSubscribed telemetry not sent
+		tm := sm.params.Telemetry.(*telemetryfakes.FakeTelemetryService)
+		require.Equal(t, 1, tm.TrackSubscribeRequestedCallCount())
+		require.Equal(t, 0, tm.TrackSubscribedCallCount())
```

Comment:
this is a good question. I think we should still treat permissions failures as failed events.. maybe logging them explicitly so that the user can be aware?

what do you think?

Reply:
Yes and no. 

yes for logging them clearly and give users the visibility.

no for treating them as failure. Saying no because system is behaving as expected, i. e. some subscription was desired, but there was no permission. For example spatial applications could enable auto-subscribe (intentionally or by mistake) and that could potentially load to lot of subscription attempts with a majority of them failing due to permissions as users are spread out in the space. My main concern is that attempts vs successes metric getting used to define health of a node. We can still count attempts. Maybe introduce a new counter like `permissionsFailed` and then do the failure percentage as `(numFailures - permissionsFailed) / numAttempts` and use that to effect any changes to node health.

wdyt? cc: @real-danm

Reply:
good question regarding what they will be used for. Perhaps the subscription itself can now generate different events:
* requested
* success
* no permission / track not found / etc -- failures that are outside of the responsibility of the subscription system
* subscription failure

Then we can choose to use these stats in different ways.

Reply:
Yeah, that would be my preference.